module/VuFind/src/VuFind/DeveloperSettings/DeveloperSettingsService.php

```diff
@@ -0,0 +1,222 @@
+     * Generate an API key for a user.
+     *
+     * @param UserEntityInterface $user  User
+     * @param string              $title Title for the API key
+     *
+     * @return ApiKeyEntityInterface|false API key entity on success, false on failure.
+     */
+    public function generateApiKeyForUser(UserEntityInterface $user, string $title): ApiKeyEntityInterface|false
+    {
+        $tokens = $this->apiKeyService->getApiKeysForUser($user);
+        if ($this->apiKeysBlocked($tokens)) {
+            return false;
+        }
+        // Generate unique id from date and users id.
+        $newKey = $this->apiKeyService->createEntity();
+        $date = new DateTime();
+        $newKey->setToken($this->createNewToken($user))
+            ->setUser($user)
+            ->setCreated($date)
+            ->setLastUsed($date)
+            ->setTitle($title);
+        $this->apiKeyService->persistEntity($newKey);
+        return $newKey;
+    }
+
+    /**
+     * Set the last used value to the API key. By default this will be only updated once every hour.
+     *
+     * @param ApiKeyEntityInterface $apiKey API key
+     *
+     * @return void
+     */
+    protected function updateLastUsed(ApiKeyEntityInterface $apiKey): void
+    {
+        if (time() - $apiKey->getLastUsed()->getTimestamp() < $this->updateInterval * 60) {
+            $apiKey->setLastUsed(new \DateTime());
+            $this->apiKeyService->persistEntity($apiKey);
+        }
+    }
+
+    /**
+     * Is the user blocked from generating new API keys?
+     *
+     * @param ApiKeyEntityInterface[] $keys Users keys
+     *
+     * @return bool
+     */
+    public function apiKeysBlocked(array $keys): bool
+    {
+        foreach ($keys as $key) {
+            if ($key->isRevoked()) {
+                return true;
+            }
+        }
+        return count($keys) >= $this->keyLimitPerUser;
+    }
+
+    /**
+     * Delete an API key for a user
+     *
+     * @param UserEntityInterface $user User
+     * @param int                 $id   API key id
+     *
+     * @return bool
+     */
+    public function deleteApiKeyForUser(UserEntityInterface $user, int $id): bool
+    {
+        $key = $this->apiKeyService->getByUserAndId(
+            $user,
+            $id
+        );
+        if (false === $key?->isRevoked()) {
+            $this->apiKeyService->deleteApiKey($key);
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Check if API keys are enabled.
+     *
+     * @return bool
+     */
+    public function apiKeysEnabled(): bool
+    {
+        return DeveloperSettingsStatus::settingEnabled($this->apiKeySettings['mode'] ?? '');
+    }
+
+    /**
+     * Check if the provided token is valid. If user provides an API key token
+     * in optional mode, it will act the same way even if
+     *
+     * @param ?string $token Token to search for API key
+     *
+     * @return bool
+     */
+    public function isTokenValid(?string $token): bool
```

Comment:
I still think it would be better to call this `isValidApiToken`. Or maybe `isApiKeyAllowed` is better since it doesn't imply the key must be valid. It's a subtle difference, but `isTokenValid` feels like it should always be passed a valid token for it to return true.

Reply:
changed it to isApiKeyAllowed